pkg/ingester/user_tsdb.go

```diff
@@ -467,3 +504,113 @@ func (u *userTSDB) releaseAppendLock(acquireState tsdbState) {
 		u.inFlightAppendsStartedBeforeForcedCompaction.Done()
 	}
 }
+
+func (u *userTSDB) ownedSeriesAndShards() (int, int) {
+	u.ownedSeriesMtx.Lock()
+	defer u.ownedSeriesMtx.Unlock()
+
+	return u.ownedSeriesCount, u.ownedSeriesShardSize
+}
+
+func (u *userTSDB) getAndClearReasonForRecomputeOwnedSeries() string {
+	return u.requiresOwnedSeriesUpdate.Swap("")
+}
+
+func (u *userTSDB) triggerRecomputeOwnedSeries(reason string) {
+	u.requiresOwnedSeriesUpdate.CompareAndSwap("", reason)
+}
+
+// recomputeOwnedSeries recomputes owned series for current token ranges, and updates both owned series and shard size.
+//
+// This method returns true, if recomputation of owned series failed multiple times due to too
+// many new series being added during the computation. If no such problem happened, this method returns false.
+//
+// This method and updateTokenRanges should be only called from the same goroutine. (ownedSeries service)
+func (u *userTSDB) recomputeOwnedSeries(shardSize int, reason string, logger log.Logger) (retry bool) {
+	retry, _ = u.recomputeOwnedSeriesWithComputeFn(shardSize, reason, logger, u.computeOwnedSeries)
+	return retry
+}
+
+const (
+	recomputeOwnedSeriesMaxAttemps    = 3
+	recomputeOwnedSeriesMaxSeriesDiff = 1000
+)
+
+func (u *userTSDB) recomputeOwnedSeriesWithComputeFn(shardSize int, reason string, logger log.Logger, compute func() int) (retry bool, _ int) {
+	start := time.Now()
+
+	prevOwnedSeriesCount, shardSizePrev := u.ownedSeriesAndShards()
+
+	reportError := false
+	attempts := 0
+	var ownedNew int
+	for attempts < recomputeOwnedSeriesMaxAttemps {
+		attempts++
+
+		ownedNew = compute()
+
+		u.ownedSeriesMtx.Lock()
+
+		// Check how many new series were added while we were computing owned series.
+		seriesDiff := u.ownedSeriesCount - prevOwnedSeriesCount
+		seriesDiffOk := seriesDiff >= 0 && seriesDiff <= recomputeOwnedSeriesMaxSeriesDiff // seriesDiff should always be >= 0, but in case it isn't, we can try again.
+		if seriesDiffOk || attempts == recomputeOwnedSeriesMaxAttemps {
+			// If less than maxSeriesDiff were added while computing owned series, we can update our values.
+			u.ownedSeriesCount = ownedNew
+			u.ownedSeriesShardSize = shardSize
+
+			u.ownedSeriesMtx.Unlock()
+
+			if !seriesDiffOk {
+				reportError = true
+			}
+			break
+		}
+
+		// If too many series were created in the meantime, our new number of owned series may be wrong
+		// (it may or may not include the new series, we don't know). In that case, just run the computation again.
+		prevOwnedSeriesCount = u.ownedSeriesCount
+		shardSizePrev = u.ownedSeriesShardSize
+
+		u.ownedSeriesMtx.Unlock()
+	}
+
+	level.Info(logger).Log("msg", "owned series: recomputed owned series for user",
```

Comment:
[non blocking] Do we really have to log it for every tenant? I guess it can help you debugging in case of issues for now, but we may consider getting rid of it once feature will be stable.

Reply:
Agree to remove it in the future.